src/util.cpp

```diff
@@ -247,38 +247,94 @@ Thread gettid()
     return Thread(syscall(SYS_gettid));
 }
 
-std::map<dev_t, std::string> get_block_devices()
+std::vector<BlockDevice> get_block_devices()
 {
-    std::map<dev_t, std::string> result;
+    std::vector<BlockDevice> result;
     std::filesystem::path sys_dev_path("/sys/dev/block");
 
-    const std::regex devname_regex("DEVNAME=(.+)");
-    const std::regex major_minor_regex("/sys/dev/block/(\\d+):(\\d+)");
+    const std::regex devname_regex("DEVNAME=(\\S+)");
+    const std::regex devtype_regex("DEVTYPE=(\\S+)");
+    const std::regex major_regex("MAJOR=(\\S+)");
+    const std::regex minor_regex("MINOR=(\\S+)");
+
     std::smatch devname_match;
-    std::smatch major_minor_match;
+    std::smatch devtype_match;
+    std::smatch major_match;
+    std::smatch minor_match;
 
     for (const std::filesystem::directory_entry& dir_entry :
          std::filesystem::directory_iterator(sys_dev_path))
     {
         std::string path_str = dir_entry.path().string();
-        std::regex_match(path_str, major_minor_match, major_minor_regex);
-        std::string devname = fmt::format("unknown device {}:{}", major_minor_match[1].str(),
-                                          major_minor_match[2].str());
+        std::string devname = "unknown device";
+        uint32_t major = 0;
+        uint32_t minor = 0;
+        std::string devtype = "partition";
 
         std::filesystem::path uevent_path = dir_entry.path() / "uevent";
         std::ifstream uevent_file(uevent_path);
+
         while (uevent_file.good())
         {
             std::string line;
             uevent_file >> line;
             if (std::regex_match(line, devname_match, devname_regex))
             {
                 devname = fmt::format("/dev/{}", devname_match[1].str());
-                break;
             }
+            else if (std::regex_match(line, devtype_match, devtype_regex))
+            {
+                devtype = devtype_match[1].str();
+            }
+            else if (std::regex_match(line, major_match, major_regex))
+            {
+                major = std::stoi(major_match[1].str());
+            }
+            else if (std::regex_match(line, minor_match, minor_regex))
+            {
+                minor = std::stoi(minor_match[1].str());
+            }
+        }
+
+        uint32_t parent_major = 0;
+        uint32_t parent_minor = 0;
+        if (devtype == "partition")
+        {
+            std::filesystem::path parent_dev("/sys");
+
+            // Because someone at Linux has a serious glue-sniffing problem these symlinks are
```

Comment:
I presume they are relative paths so that someone with an even more serious glue-sniffing problem can (bind) mount `sysfs` to a different path than `/sys`.